### PR TITLE
BUG: DataFrame.agg and apply with 'size' returns a scalar

### LINKED ISSUE #39934: BUG: DataFrame.agg and apply with 'size' returns a scalar
Compare

```
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
print(df.apply('size'))
print(df.apply(['size']))
print('---')
print(df.agg('size'))
print(df.agg(['size']))
```

gives

```
6
      a  b
size  3  3
---
6
      a  b
size  3  3
```

agg/apply should operate on rows/columns.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #39934
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/apply.py [hunk 1]
```diff
@@ -414,17 +418,25 @@ def maybe_apply_str(self) -> Optional[FrameOrSeriesUnion]:
         f = self.f
         if not isinstance(f, str):
             return None
+
+        obj = self.obj
+
+        if f == "size" and isinstance(obj, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/apply.py:424]: doesn't this affect Series as well?

**Review by jreback — CHANGES_REQUESTED** [review]
rebase as well.

It might be better to detect this argument 'size' and do a translation in 
```pandas/core/base.py:    _builtin_table```

**Comment by rhshadrach** [comment]
@jreback - I merged master.

> It might be better to detect this argument 'size' and do a translation in 
```pandas/core/base.py:    _builtin_table```

This isn't a builtin; `DataFrame.size` reports the number of cells in the frame (rows * columns). Also not sure what we'd map it to, count is similar and operates on rows/columns but drops nans whereas size does not.

File: pandas/core/apply.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/apply.py:551]: For a Series, size gives the number of rows and so the default path calling `obj.size` produces the correct result.

**Comment by jreback** [comment]
my point is to map this to a lambda function to avoid special casing like this

**Comment by rhshadrach** [comment]
@jreback - makes sense, and I agree your suggestion is the prefered way. The one issue is that when replacing `func` with a lambda, the resulting Series has name None whereas tests currently expect "size". I see two ways to handle (hopefully not overlooking something simple):

 1. Mark as TODO and tackle when consolidating code for wrapping results
 2. Add a "result_name" attribute the the Apply class, set to None in all cases except this one where it is "size", and apply it in various methods where appropriate.

It seems easier and less error-prone to go with 1, but I can do 2 here if you prefer.

**Comment by jreback** [comment]
> @jreback - makes sense, and I agree your suggestion is the prefered way. The one issue is that when replacing `func` with a lambda, the resulting Series has name None whereas tests currently expect "size". I see two ways to handle (hopefully not overlooking something simple):
> 
> 1. Mark as TODO and tackle when consolidating code for wrapping results
> 2. Add a "result_name" attribute the the Apply class, set to None in all cases except this one where it is "size", and apply it in various methods where appropriate.
> 
> It seems easier and less error-prone to go with 1, but I can do 2 here if you prefer.

yeah happy to do this in the future. 2 sounds interesting actually. why don't you add a TODO for now (and issue) then can merge this one.

**Comment by rhshadrach** [comment]
TODO added, issue opened, and green(ish - failure is unrelated).

**Comment by jreback** [comment]
can you merge master & ping on greenish

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rhshadrach

File: pandas/core/apply.py [hunk 2]
```diff
@@ -541,17 +545,26 @@ def maybe_apply_str(self) -> Optional[FrameOrSeriesUnion]:
...
         if not isinstance(f, str):
             return None
+
+        obj = self.obj
+
+        # TODO: GH 39993 - Avoid special-casing by replacing with lambda
+        if f == "size" and isinstance(obj, ABCDataFrame):
+            # Special-cased because DataFrame.size returns a single scalar
+            value = obj.shape[self.axis]
+            return obj._constructor_sliced(value, index=self.agg_axis, name="size")
```
Comment (TheNeuralBit) [inline pandas/core/apply.py:555]: Is it intentional to set the name here? This isn't consistent with other string agg methods

Comment (rhshadrach) [reply pandas/core/apply.py:555]: Indeed, thanks @TheNeuralBit! Opened #42046

Comment (TheNeuralBit) [reply pandas/core/apply.py:555]: Thank you!

**Pull request merged**